File: projects/algorithms-and-leetcode/leetcode/150_top_problems/c++/best_time_to_buy_and_sell_stockII.cpp
// You are given an integer array prices where prices[i] is the price of a given stock on the ith day.

// On each day, you may decide to buy and/or sell the stock. You can only hold at most one share of the stock at any time. However, you can buy it then immediately sell it on the same day.

// Find and return the maximum profit you can achieve.

#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int max_profit = 0;
        for (int i = 1; i < static_cast<int>(prices.size()); ++i) {
            if (prices[i] > prices[i - 1]) {
                max_profit += prices[i] - prices[i - 1];
            }
        }
        return max_profit;
    }
};

int main() {
    Solution solution;
    vector<int> prices = {7, 1, 5, 3, 6, 4};
    int result = solution.maxProfit(prices);
    cout << "Maximum Profit: " << result << endl; // Expected: 7
    return 0;
}

File: projects/algorithms-and-leetcode/leetcode/150_top_problems/python/best_time_to_buy_and_sell_stockII.py
"""
You are given an integer array prices where prices[i] is the price of a given stock on the ith day.

On each day, you may decide to buy and/or sell the stock. You can only hold at most one share of the stock at any time. However, you can buy it then immediately sell it on the same day.

Find and return the maximum profit you can achieve.

"""
from typing import List

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        """This is a DP programming problem."""
        max_profit = 0
        for i in range(1, len(prices)):
            #print(i)
            if prices[i] > prices[i - 1]:
                #print(prices[i], prices[i - 1])
                max_profit += prices[i] -  prices[i -1]
                print(max_profit)
        return max_profit
    


if __name__ == "__main__":
    prices = [7, 1, 5, 3, 6, 4]
    solution = Solution()
    print(solution.maxProfit(prices))  # Output: 7
    # prices = [1, 2, 3, 4, 5]
    # print(solution.maxProfit(prices)) # Output: 4